share/steel/examples/pulse/lib/Pulse.Lib.LinkedList.fst

```diff
@@ -399,59 +382,59 @@ fn is_last_cell (#t:Type) (x:llist t)
     returns b:bool
     ensures is_list x 'l ** pure (b == (List.Tot.length 'l = 1))
 {
-    let np = Some?.v x;
-    is_list_cases_some x np;
-    let node = !np;
-    with tail tl. assert (is_list #t tail tl);
-    rewrite each tail as node.tail;
-    match node.tail {
-        None -> { 
-            is_list_cases_none node.tail;
-            intro_is_list_cons x np;
-            true
-        }
-        Some vtl -> { 
-            is_list_cases_some node.tail vtl;
-            intro_is_list_cons node.tail vtl;
-            intro_is_list_cons x np;
-            false
-        }
+  let np = Some?.v x;
+  is_list_cases_some x np;
+  let node = !np;
+  with tail tl. assert (is_list #t tail tl);
+  rewrite each tail as node.tail;
+  match node.tail {
+    None -> { 
+      is_list_cases_none node.tail;
+      intro_is_list_cons x np;
+      true
+    }
+    Some vtl -> { 
+      is_list_cases_some node.tail vtl;
+      intro_is_list_cons node.tail vtl;
+      intro_is_list_cons x np;
+      false
     }
+  }
 }
 ```
 
+#push-options " --query_stats --log_queries" // --debug Pulse.Lib.LinkedList --debug_level SMTQuery  --log_queries"
+#restart-solver
+
+
 ```pulse
 fn append_at_last_cell (#t:Type) (x y:llist t)
-    requires
-        is_list x 'l1 **
-        is_list y 'l2 **
-        pure (Some? x /\ List.Tot.length 'l1 == 1)
-    ensures
-        is_list x (List.Tot.append 'l1 'l2)
+requires
+  is_list x 'l1 **
+  is_list y 'l2 **
+  pure (Some? x /\ List.Tot.length 'l1 == 1)
+ensures
+  is_list x (List.Tot.append 'l1 'l2)
 {
-    let np = Some?.v x;
-    is_list_cases_some x np;
-    let node = !np;
-    with tail tl. assert (is_list #t tail tl);
-    rewrite each tail as node.tail;
-    match node.tail {
-        None -> {
-            is_list_cases_none node.tail;
-            elim_is_list_nil node.tail;
-            np := mk_node node.head y;
-            rewrite each y as (mk_node node.head y).tail in (is_list y 'l2);
-            intro_is_list_cons x np; 
-        }
-        Some vtl -> {
-            //need an easier way to do elim_false
-            is_list_cases_some node.tail vtl;
-            intro_is_list_cons node.tail vtl;
-            intro_is_list_cons x np;
-            elim_false (is_list x (List.Tot.append 'l1 'l2));
-            drop (is_list x 'l1);
-            drop (is_list y 'l2)
-        }
+  let np = Some?.v x;
+  is_list_cases_some x np;
+  let node = !np;
+  with tail tl. assert (is_list #t tail tl);
+  rewrite each tail as node.tail;
+  match node.tail {
+    None -> {
+      is_list_cases_none node.tail;
+      elim_is_list_nil node.tail;
+      np := mk_node node.head y;
+      rewrite each y as (mk_node node.head y).tail in (is_list y 'l2);
+      intro_is_list_cons x np; 
+    }
+    Some vtl -> {
+      is_list_cases_some node.tail vtl;
+      assert (pure False); //adding this somehow seems necessary; weird
+      unreachable ();
```

Comment:
A little bit of SMT strangeness here: I needed to explicitly assert False first and only then could I prove unreachable ... to be investigated.